Bottom: f150652ca82a5866f82c590caaa5312f4a3a62bf
Top:    e5d2faf273bb5e241d48185f994338b93c1b1a76
Author: Ian Kent <ikent@redhat.com>
Date:   2013-04-04 18:16:28 +0800

autofs-5.0.7 - syncronize handle_mounts() shutdown-2


---

diff --git a/daemon/automount.c b/daemon/automount.c
index 4c651cf..96a55b3 100644
--- a/daemon/automount.c
+++ b/daemon/automount.c
@@ -1449,6 +1449,7 @@ static void handle_mounts_cleanup(void *arg)
 	char path[PATH_MAX + 1];
 	char buf[MAX_ERR_BUF];
 	unsigned int clean = 0, submount, logopt;
+	unsigned int pending = 0;
 
 	ap = (struct autofs_point *) arg;
 
@@ -1466,6 +1467,9 @@ static void handle_mounts_cleanup(void *arg)
 		list_del_init(&ap->mounts);
 	}
 
+	/* Don't signal the handler if we have already done so */
+	if (!list_empty(&master_list->completed))
+		pending = 1;
 	master_remove_mapent(ap->entry);
 	master_source_unlock(ap->entry);
 
@@ -1498,7 +1502,7 @@ static void handle_mounts_cleanup(void *arg)
 	 * so it can join with any completed handle_mounts() threads and
 	 * perform final cleanup.
 	 */
-	if (!submount)
+	if (!submount && !pending)
 		pthread_kill(state_mach_thid, SIGTERM);
 
 	master_mutex_unlock();
